Expose display’s root_group, add function to resize REPL terminal

kmatch98 opened the pull request:
This exposes the REPL's display group elements as `supervisor.splash`.  Also, this adds a function to resize the dimensions of the REPL terminal window.  This allows the user to view the REPL and console `print()` output as well as view any display elements at the same time.

**Warning**: This is somewhat dangerous since exposing the REPL display elements allows the the REPL to be relocated where it is no longer visible.  

Here is an example creating a split screen with the REPL at the bottom of the display and displayio elements at the top of the screen.

https://user-images.githubusercontent.com/33587466/151874447-ea85c1cc-e0d5-43df-b2c7-e8fbaccb3a1f.MOV

``` python
import board
import displayio
import supervisor
import time


display=board.DISPLAY

# Create a bitmap with two colors
bitmap1 = displayio.Bitmap(display.width//4, display.height//4 - 10, 2)

# Create a two color palette
palette1 = displayio.Palette(2)
palette1[0] = 0xff00ff
palette1[1] = 0xffffff

# Create a TileGrid using the Bitmap and Palette
tile_grid1 = displayio.TileGrid(bitmap1, pixel_shader=palette1)


# Create a bitmap with two colors
bitmap2 = displayio.Bitmap(display.width//4, display.height//4 - 10, 2)

# Create a two color palette
palette2 = displayio.Palette(2)
palette2[0] = 0x00ffff
palette2[1] = 0xffffff

# Create a TileGrid using the Bitmap and Palette
tile_grid2 = displayio.TileGrid(bitmap2, pixel_shader=palette2)
tile_grid2.x=display.width//4

# Create a bitmap with two colors
bitmap3 = displayio.Bitmap(display.width//4, display.height//4 - 10, 2)

# Create a two color palette
palette3 = displayio.Palette(2)
palette3[0] = 0xffff00
palette3[1] = 0xffffff

# Create a TileGrid using the Bitmap and Palette
tile_grid3 = displayio.TileGrid(bitmap3, pixel_shader=palette3)
tile_grid3.x=2*display.width//4

# Create a Group
mygroup = displayio.Group()

print()
print()
print()
print("REPL control test: Taking control of REPL group")
time.sleep(3)

# Note: You must "display.show" your own group before adding the supervisor.splash to your own group.
# Reason: When displaying the normal REPL (for example with display.show(None), the supervisor.splash
# group is already in a group that is displayed.  To remove the supervisor.splash from the displayed group,
# you first have to display.show some other group, doing that will remove the supervisor.splash from its group
# and allow you to append it to your own group.

display.show(mygroup)
# add the supervisor.splash group to the displayed group.
mygroup.append(supervisor.splash)

# resize the supervisor.splash group pixel dimensions, make it half the display height.
supervisor.reset_terminal(display.width, display.height//2)

# relocate the supervisor.splash group on the display, moving it half-way down the display
supervisor.splash.y=display.height//2
print("Resize and move the splash screen")
time.sleep(2)

# demonstrate how print statements scroll on the REPL/console display
print("Add some prints to show terminal scrolling:")
time.sleep(0.3)
for i in range(6):
	print("Line: {}".format(i))
	time.sleep(0.3)
time.sleep(1.5)

# Add bitmap rectangles to the display group
print("Add some displayio bitmap rectangles:")
time.sleep(1)

# Add the TileGrid to the Group
print("Add a pink rect.")
mygroup.append(tile_grid1)
time.sleep(1)
print("Add a cyan rect.")
mygroup.append(tile_grid2)
time.sleep(1)
print("Add a yellow rect.")
mygroup.append(tile_grid3)
time.sleep(1)

# print some more lines to show the console text scrolling, with 
# the bitmaps still up on the upper half of the display
for i in range(6):
	print("Another line feed: {}!".format(i))
	time.sleep(0.3)
time.sleep(3)
print("Resetting the REPL and ending the code")
time.sleep(1)

# Before finishing this code, set the terminal size to the full display pixel dimensions
supervisor.reset_terminal(display.width, display.height)
```

tannewt reviewed (commented):
I like this functionality but I've got some questions:

* Could we add the ability to get the current Group from Display instead of having a separate accessor?
* Could we reset the location of the terminal at a certain point in case things go awry?
* Would it be more versatile to provide the serial input to terminal into userland so that it can create it's own Terminal?

anecdata commented:
This is a cool feature, and probably faster than approximating it with a partial-screen `terminalio.Terminal`(which also probably needs some gyrations to get print() statements to be natural).

kmatch98 commented:
Thanks for the constructive comments. 

- being able to access the current Group would be more general. Yet if you’re somewhere already deep in your display code, it would require you to do a `display.show(None)` to access a handle to the splash group. Probably not a huge deal and you could turn off auto_refresh so it would be hidden to the user. Let me know your decision and I’ll make edits. 
- I assume that when code.py stops that it essentially does a `display.show(None)`. I’ll hunt for where the `display.show(None)` happens and that can reset the display to its original position and size. 
- I’m with @anecdata on this one. A user can build their own terminal, but if the REPL is already there, that makes it a lot simpler to use. 
- One other idea I toyed with is whether resizing the REPL should also forward over any text data from the previous incarnation of the Terminal tilegrid. Currently the tilegrid is obliterated and a fresh one is created. However that would require allocating the new tilegrid memory, copying the tilegrid data that will fit and then deallocating the previous tilegrid. Is it worth this effort/code space to keep the REPL text even when the size changes?

kmatch98 commented:
I totally messed up my commits. I’ll see if I can enlist help on discord to get the revision sorted out.

kmatch98 commented:
Thanks @FoamyGuy for the good feedback. Per @tannewt comments above I modified the functionality and naming.  To clarify, here is the current status of what is added and the names. 

- `display.root_group`: the root displayio Group that is currently displayed. To access the “splash” screen group (with Blinka and the REPL terminal output) you first have to do `display.show(None)` and then procure the `display.root_group`. If you don’t want that to disturb the display you can turn auto_refresh to false.
- `supervisor.reset_terminal(x_pixels, y_pixels)`: function to resize the REPL terminal. 
- finally, I added some code to reset the REPL to its default position and size whenever `display.show(None)` is called. This ensures that the REPL returns to its expected form whenever code.py ends.

tannewt commented:
> * One other idea I toyed with is whether resizing the REPL should also forward over any text data from the previous incarnation of the Terminal tilegrid. Currently the tilegrid is obliterated and a fresh one is created. However that would require allocating the new tilegrid memory, copying the tilegrid data that will fit and then deallocating the previous tilegrid. Is it worth this effort/code space to keep the REPL text even when the size changes?

I think this would be way more work than it's worth. The moment you change the width, you'd then need to rewrap all of the text. Nevermind the VT100 commands that can actually navigate the grid and modify it. Obliterate is my preference.

tannewt commented on shared-bindings/supervisor/__init__.c:
```diff
@@ -299,6 +300,17 @@ STATIC mp_obj_t supervisor_disable_ble_workflow(void) {
 MP_DEFINE_CONST_FUN_OBJ_0(supervisor_disable_ble_workflow_obj, supervisor_disable_ble_workflow);
 
+//| def reset_terminal(x_pixels: int, y_pixels: int) -> None:
+//|     """Adjust the pixel dimensions of the REPL console."""
```
```suggestion
//|     """Reset the CircuitPython serial terminal with new dimensions."""
```

kmatch98 replied:
Accepted changes.

tannewt commented on shared-bindings/supervisor/__init__.c:
```diff
@@ -299,6 +300,17 @@ STATIC mp_obj_t supervisor_disable_ble_workflow(void) {
+//|     ...
+//|
+STATIC mp_obj_t supervisor_reset_terminal(mp_obj_t x_pixels, mp_obj_t y_pixels) {
+    supervisor_stop_terminal();
+    supervisor_start_terminal(mp_obj_get_int(x_pixels), mp_obj_get_int(y_pixels));
```
```suggestion
    #if CIRCUITPY_DISPLAYIO
    supervisor_stop_terminal();
    supervisor_start_terminal(mp_obj_get_int(x_pixels), mp_obj_get_int(y_pixels));
    #else
    mp_raise_NotImplementedError(NULL);
    #endif
```

kmatch98 replied:
Accepted changes.

tannewt reviewed (changes requested):
Thank you for the iteration. It looks good!

The builds are failing because `supervisor_stop_terminal` is only linked in for boards that support displayio. I've added a suggestion that should fix it.

I agree the term "splash" isn't ideal but that's what I used when I added it. Maybe "serial_terminal" is clearer? I don't care if you change it or not in this PR. The term is only used internally so we're not committing to it.

FoamyGuy reviewed (commented):
I attempted to build from this branch and run on a PyPortal but it seems that I'm getting a hard crash.

Immediately after the UF2 finishes pasting, the device reboots as normal. And for just one second I can see the serial output on the display as normal: 

![pyportal_troubleshoot](https://user-images.githubusercontent.com/2406189/152079658-61000cf3-ef68-4096-8b83-7b4c84eb85bf.png)

But immediately after this the display goes dark and I never see `CIRCUITPY` or the serial port connect from the PC side. It only shows this very briefly after pasting UF2 and only once. After it disappears the first time then it doesn't ever get shown again until changing back to `7.1.1` and then back to the version from this branch.

On subsequent resets from the button and unplug/replug nothing happens on the display it stays dark. But the onboard LEDs do seem blink as normal.

kmatch98 commented:
@FoamyGuy this pyportal issue was resolved by
https://github.com/adafruit/circuitpython/pull/5951

I pulled the repo before that PR was merged so it’s not incorporated in this repo. Do I need to merge in those changes?

FoamyGuy commented:
@kmatch98 ah I see! I don't think you need to merge it to here(I believe github works out the differences correctly in most cases), but I am not certain. I'll try it again tomorrow and merge that PR into my local repo to build.

kmatch98 commented:
Fixed an erroneous documentation string. Local docs build ok now.  This is ready for another review.

tannewt commented:
@dhalbert did you reduce the `datum_imu` board size already?

tannewt reviewed (approved):
This looks good to me! I think the one board that's too big may be fixed by another PR so we can merge this.

dhalbert commented:
> @dhalbert did you reduce the `datum_imu` board size already?

Yes, fixed by https://github.com/adafruit/circuitpython/pull/5969/commits/0d43e3e734e52c262323c1d509b332f38874b76d

kmatch98 commented:
Here's the latest demo code for moving the REPL position and adding some displayio objects:

``` python
import board
import displayio
import supervisor
import time


display=board.DISPLAY

# Create a bitmap with two colors
bitmap1 = displayio.Bitmap(display.width//4, display.height//4 - 10, 2)

# Create a two color palette
palette1 = displayio.Palette(2)
palette1[0] = 0xff00ff
palette1[1] = 0xffffff

# Create a TileGrid using the Bitmap and Palette
tile_grid1 = displayio.TileGrid(bitmap1, pixel_shader=palette1)


# Create a bitmap with two colors
bitmap2 = displayio.Bitmap(display.width//4, display.height//4 - 10, 2)

# Create a two color palette
palette2 = displayio.Palette(2)
palette2[0] = 0x00ffff
palette2[1] = 0xffffff

# Create a TileGrid using the Bitmap and Palette
tile_grid2 = displayio.TileGrid(bitmap2, pixel_shader=palette2)
tile_grid2.x=display.width//4

# Create a bitmap with two colors
bitmap3 = displayio.Bitmap(display.width//4, display.height//4 - 10, 2)

# Create a two color palette
palette3 = displayio.Palette(2)
palette3[0] = 0xffff00
palette3[1] = 0xffffff

# Create a TileGrid using the Bitmap and Palette
tile_grid3 = displayio.TileGrid(bitmap3, pixel_shader=palette3)
tile_grid3.x=2*display.width//4

# Create a Group
mygroup = displayio.Group()

print()
print()
print()
print("REPL control test: Taking control of REPL group")
time.sleep(3)


# clear the display to the REPL 
display.show(None)
splash = board.DISPLAY.root_group # this gets the current root_group, the REPL

# Note: You must "display.show" your own group before adding the splash to your own group.
# Reason: When displaying the normal REPL (for example with display.show(None), the splash
# group is already in a group that is displayed.  To remove the splash from the displayed group,
# you first have to display.show some other group, doing that will remove the splash from its group
# and allow you to append it to your own group.	
display.show(mygroup)

# resize the supervisor.splash group pixel dimensions, make it half the display height.
supervisor.reset_terminal(display.width, display.height//2)

# relocate the supervisor.splash group on the display, moving it half-way down the display
splash.y=display.height//2
print("Resize and move the splash screen")

# append the supervisor.splash group to the displayed group.
mygroup.append(splash)
time.sleep(2)

# demonstrate how print statements scroll on the REPL/console display
print("Add some prints to show terminal scrolling:")
time.sleep(0.3)
for i in range(6):
	print("Line: {}".format(i))
	time.sleep(0.3)
time.sleep(1.5)

# Add bitmap rectangles to the display group
print("Add some displayio bitmap rectangles:")
time.sleep(1)

# Add the TileGrid to the Group
print("Add a pink rect.")
mygroup.append(tile_grid1)
time.sleep(1)
print("Add a cyan rect.")
mygroup.append(tile_grid2)
time.sleep(1)
print("Add a yellow rect.")
mygroup.append(tile_grid3)
time.sleep(1)

# print some more lines to show the console text scrolling, with 
# the bitmaps shown on the upper half of the display
for i in range(6):
	print("Another line feed: {}!".format(i))
	time.sleep(0.3)
time.sleep(3)
print("Ending the code")
time.sleep(1)
```

matemaciek commented:
> ```python
> splash.y=display.height//2
> ```

On CP 9.0.0 this line results in `RuntimeError: Read-only`. Is there any other way of shifting the terminal on display now?

jepler commented:
@matemaciek Please file a new PR about this message. This most likely is an undesired side-effect of https://github.com/adafruit/circuitpython/pull/8923 that should be fixed.
